bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/GatewayWebTargets.java

```diff
@@ -0,0 +1,450 @@
+            invoke(HttpMethod.PUT, register, null, null);
+            registered = true;
+        }
+        return registered;
+    }
+
+    /**
+     * Get hub properties.
+     *
+     * @return a map containing the hub properties.
+     * @throws HubProcessingException if any error occurs.
+     */
+    public Map<String, String> getInformation() throws HubProcessingException {
+        String json = invoke(HttpMethod.GET, info, null, null);
+        try {
+            Info result = jsonParser.fromJson(json, Info.class);
+            if (result == null) {
+                throw new HubProcessingException("getInformation(): missing response");
+            }
+            return Map.of( //
+                    Thing.PROPERTY_FIRMWARE_VERSION, result.getFwVersion(), //
+                    Thing.PROPERTY_SERIAL_NUMBER, result.getSerialNumber());
+        } catch (JsonParseException e) {
+            throw new HubProcessingException("getFirmwareVersions(): JsonParseException");
+        }
+    }
+
+    /**
+     * Get the list of scenes.
+     *
+     * @return the list of scenes.
+     * @throws HubProcessingException if any error occurs.
+     */
+    public List<Scene> getScenes() throws HubProcessingException {
+        String json = invoke(HttpMethod.GET, scenes, null, null);
+        try {
+            Scene[] scenes = jsonParser.fromJson(json, Scene[].class);
+            if (scenes == null || Arrays.stream(scenes).anyMatch(Objects::isNull)) {
+                throw new HubProcessingException("getScenes() response is null or contains null(s)");
+            }
+            return List.of(scenes);
+        } catch (JsonParseException e) {
+            throw new HubProcessingException("getScenes() JsonParseException");
+        }
+    }
+
+    /**
+     * Get the data for a single shade.
+     *
+     * @param shadeId the id of the shade to get.
+     * @return the shade.
+     * @throws HubProcessingException if any error occurs.
+     */
+    public Shade getShade(int shadeId) throws HubProcessingException {
+        String json = invoke(HttpMethod.GET, String.format(shadeSingle, shadeId), null, null);
+        try {
+            Shade result = jsonParser.fromJson(json, Shade.class);
+            if (result == null) {
+                throw new HubProcessingException("getShade() missing response");
+            }
+            return result;
+        } catch (JsonParseException e) {
+            throw new HubProcessingException("getShade() JsonParseException");
+        }
+    }
+
+    /**
+     * Get the list of shades.
+     *
+     * @return the list of shades.
+     * @throws HubProcessingException if any error occurs.
+     */
+    public List<Shade> getShades() throws HubProcessingException {
+        String json = invoke(HttpMethod.GET, shades, null, null);
+        try {
+            Shade[] shades = jsonParser.fromJson(json, Shade[].class);
+            if (shades == null || Arrays.stream(shades).anyMatch(Objects::isNull)) {
```

Comment:
`fromJson` returns only null if provided with null or empty json. This can happen only for HTTP code 204 as `invoke` will otherwise throw in this case:
```java
        if ((HttpStatus.OK_200 == statusCode) && ((jsonResponse == null) || (jsonResponse.isEmpty()))) {
            throw new HubProcessingException("Missing response entity");
        }
```

But since you're throwing anyway here, I'm wondering why you want `invoke` to allow returning null or empty json for 204 rather than just throwing also in this case? It seems to just transfer this problem to the caller instead of dealing directly with it. Otherwise could be be simplified to:
```suggestion
            if (Arrays.stream(shades).anyMatch(Objects::isNull)) {
```

I have not checked if objects can be null, but it wouldn't make much sense, I guess at least empty objects should be returned.

Reply:
Or return `List.of()` here instead of throwing - if 204 is expected and should be interpreted as "no shades" rather than an error.

Reply:
> can happen only for HTTP code 204

Ignore this special case. The gateway only returns 204 for the integration registration..

![image](https://github.com/openhab/openhab-addons/assets/893994/b0f0e241-c43a-4955-991b-ed0d73bf90b6)

Reply:
^
I will revert the change and let the NPE (which as you imagine) would never happen, simply escalate (if it would ever so do)..

Reply:
Wouldn't the registration then go wrong? Perhaps 204 should only be allowed for `PUT`, but otherwise throw?

Reply:
Just to be clear: I think in addition to reverting the checks, this would make the error handling safer (at the end of `invoke`):
```java
        if (method == HttpMethod.GET && (jsonResponse == null || jsonResponse.isEmpty())) {
            throw new HubProcessingException("Missing response entity");
        }
```

Reply:
Ok, although I will remove the 'jsonResponse == null' case which cannot occur because 'response.getContentAsString()' cannot return 'null'.